crates/hotshot/src/tasks/mod.rs

```diff
@@ -101,42 +102,57 @@ pub fn add_network_message_task<
 
     let network = Arc::clone(channel);
     let mut state = network_state.clone();
+    let shutdown_signal = Arc::clone(&handle.shutdown_signal);
     let task_handle = async_spawn(async move {
+        let mut shutdown = Box::pin(shutdown_signal.wait_for_shutdown().fuse());
         loop {
-            let decided_upgrade_certificate_lock = decided_upgrade_certificate.read().await.clone();
-            let msgs = match network.recv_msgs().await {
-                Ok(msgs) => {
-                    let mut deserialized_messages = Vec::new();
-
-                    for msg in msgs {
-                        let deserialized_message = match VersionedMessage::deserialize(
-                            &msg,
-                            &decided_upgrade_certificate_lock,
-                        ) {
-                            Ok(deserialized) => deserialized,
-                            Err(e) => {
-                                tracing::error!("Failed to deserialize message: {}", e);
-                                continue;
+            if shutdown_signal.is_shutdown() {
+                tracing::error!("Shutting down network message task !!!");
+                break;
+            }
+            futures::select! {
+                    _ = shutdown => {
+                        tracing::error!("Shutting down network message task !!!");
+                        return;
+                    }
```

Comment:
It looks like in this particular case, we have to handle the signal event twice, could we not just keep this in the future select?